### PR TITLE
Rails 4.2 upgrade

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
~~This PR is now pointing at `rails-4-1-api` which combines `rails-4-1` and [these changes](https://github.com/openfoodfoundation/openfoodnetwork/issues/6343) to API controllers and strong params.~~ There are 2 PRs marked TEMP which include some changes that are in the pipe but not merged yet. They'll disappear after merging and rebasing.

#### ActiveJob

`ActiveJob` is now a core part of Rails, essentially it's a wrapper for the interfaces of various job-processing services, like `DelayedJob`, `Sidekiq`, or `Resque`. Instead of passing a job directly to the service, it's now passed to `ActiveJob`, which passes it to the specified backend. This means the backend can be seamlessly switched out for a different one with one line of code, or in testing it can be substituted for a mock backend that is easier to work with.

`ActionMailer` has also had various changes which are aimed at integrating it with `ActiveJob`.

#### Responders

A large amount of functionality related to how controllers respond to requests has been extracted out from Rails core to the `responders` gem. This includes lots of methods like `#responds_to` and `#respond_with`. This move comes with a lot of associated changes and refactoring behind the scenes in base classes like `ActionController::Base`.

#### StateMachine -> StateMachines

The `state_machine` gem was abandoned by it's maintainers and is incompatible with Rails 4.2. It was forked and now has a fixed and maintained alternative called `state_machines`.

#### Paranoia

The `paranoia` gem that handles soft-deleting has some problems with Rspec transactions in Rails 4.2, creating lots of broken specs around soft-deletion. This bug is patched in the `rails-4` branch in their repo, but that branch is totally out of date and not available in any releases. :see_no_evil: I've applied their patch manually in an initializer and it's working as intended.

#### What to test
Added by @luisramos0
On top of carefully testing everything. These are some of the details we may look at (places where the code has been adapted).

From the rails 4.1 PR:
- xero_invoices_report - amounts and included tax
- sales_tax_report - included tax
- payments_report.rb - payments and orders amounts
- customers_report.rb - sanity check
- line item quantities in:
- - orders_and_fulfillments_report/supplier_totals_report.rb
- - orders_and_fulfillments_report/supplier_totals_by_distributor_report.rb
- - orders_and_fulfillments_report/distributor_totals_by_supplier_report.rb
- - orders_and_fulfillments_report/default_report.rb
- - orders_and_fulfillments_report/customer_totals_report.rb
- - group_buy_report.rb
- state machine - verify orders workflow in the backoffice - create order and check state address, add customer details and check state payment, add payment and check state complete (this was also touched in the rails 4.2 work)
- Included tax vs excluded tax in order adjustments like shipping AND Included tax vs excluded tax in order line items - I think this can be validated by placing orders with different tax setups (tax rates and tax categories) and see if the tax values are displayed correctly in the backoffice after the order is placed.

From this PR:
- enterprises bulk update - when you are an admin go to list of enterprises and try to update some of the fields - I think this broken in master now...
- Delayed Job: bulk invoice printing (the email and the content of the invoice) and subscription placement and confirmation
- we need to verify all emails in the system
- soft-deletion of payment methods, products, tax categories, tax rates and variants
- updating multiple records via collections / ModelSet

### CODE REVIEW

**Review by luisramos0 — COMMENTED** [review]
Awesome, we just need some serious testing notes :tada:

**Comment by Matt-Yorkley** [comment]
I think the main bits to look at are: 
- anything that goes through Delayed Job
- all emails
- soft-deletion
- updating records via collections / `ModelSet`

**Comment by luisramos0** [comment]
yes, that's it. But I think we need a list of pages for the testers to know where to test: where was delayed job changed (pdf generation and what else?), what pages are influenced by the changes in modelset, what pages are influenced by the changes in strong params?

**Comment by Matt-Yorkley** [comment]
Have we done tha for 4.1?

File: engines/order_management/app/services/order_management/stock/package.rb
Comment (luisramos0) [inline engines/order_management/app/services/order_management/stock/package.rb:104]: looking at this again, it looks broken:
image

**Comment by luisramos0** [comment]
I went for it Matt. I did it for rails 4-1 and rails 4.2 I added to the description of this PR.
The list you made for this Pr was basically it :+1:

File: engines/order_management/app/services/order_management/stock/package.rb
Comment (Matt-Yorkley) [reply engines/order_management/app/services/order_management/stock/package.rb:104]: This change wasn't actually related to an error in the build, but when I tried debugging a package object in the console with `#inspect` I got fatal errors due to "modifying a frozen string". Here we have the string `out = "#{order} - "` and then add to the string with `out <<`, but we're using **frozen string literals** in this class, so this operation is not valid, right?

Comment (Matt-Yorkley) [reply engines/order_management/app/services/order_management/stock/package.rb:104]: But yes, it looks like we need to tweak this a bit.

Comment (Matt-Yorkley) [reply engines/order_management/app/services/order_management/stock/package.rb:104]: Fix and regression test added :+1:

Comment (luisramos0) [reply engines/order_management/app/services/order_management/stock/package.rb:104]: ah, yes, just remove the last line!
thanks :+1:

**Comment by Matt-Yorkley** [comment]
Okay, I think Rails 4.2 is working with `ActionController::Metal` now...

File: app/controllers/api/base_controller.rb
Comment (luisramos0) [inline app/controllers/api/base_controller.rb:12]: great!

File: Gemfile
Comment (luisramos0) [inline Gemfile:28]: this update needs a testing note to be manually checked. currencies and potential price rounding issues?

File: app/controllers/application_controller.rb
Comment (luisramos0) [inline app/controllers/application_controller.rb:24]: this can be removed now?

File: engines/order_management/spec/services/order_management/stock/package_spec.rb
Comment (luisramos0) [inline engines/order_management/spec/services/order_management/stock/package_spec.rb:116]: typo

File: lib/application_responder.rb
Comment (luisramos0) [inline lib/application_responder.rb:9]: do we need this comment and commented line. maybe delete?

**Review by luisramos0 — APPROVED** [review]
Approved!

This needs to wait for the other PRs from master and be rebased afterwards.

File: Gemfile
Comment (Matt-Yorkley) [reply Gemfile:28]: There are lots of problems if we bump to `6.1.0` or higher. It's a cap on updates, rather than an update. It's tracked in this issue: https://github.com/openfoodfoundation/openfoodnetwork/issues/6333#issuecomment-723059261

File: app/controllers/application_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/application_controller.rb:24]: It's pretty harmless, but I think I would wait until we test and merge of the branch and remove it afterwards. It'll be useful if we encounter any issues before merging.

File: lib/application_responder.rb
Comment (Matt-Yorkley) [reply lib/application_responder.rb:9]: This file was automatically created by a generator for the `responders` gem. I assume we won't need this bit... :ok_hand:

File: app/controllers/application_controller.rb
Comment (luisramos0) [reply app/controllers/application_controller.rb:24]: ok :+1:

File: Gemfile
Comment (luisramos0) [reply Gemfile:28]: it's actually downgrading it!?
image

maybe we can keep using 5.1.1?

re < 6, I would add a short comment to the Gemfile to explain that lock, so that the next dev looking at this will not have to look around to find out.

**Comment by luisramos0** [comment]
ok, I rebased rails-4-1 on top of master (it has all we need) and then rebased this on top of rails-4-1.
Build is green :tada: 

I went through the code again, all good.
My comment about the money gem is not a major thing but I think it deserves attention before moving to test ready.

I think it would be good to have another dev look at this code before moving to Test Ready? @andrewpbrett could you have a look?

File: Gemfile
Comment (andrewpbrett) [reply Gemfile:28]: It looks like specifying `5.1.1` now conflicts with other dependencies, for `spree_i18n`. @Matt-Yorkley suggested in Slack to pull that code into the codebase and remove the gem. I think keeping the `< 6.1.0`, with a comment, and carefully testing features to do with money/rounding is the way forward for now (and maybe taking a look at the changelog for the money gem from `5.0` to `5.1.1`.

File: engines/order_management/spec/services/order_management/stock/packer_spec.rb
Comment (andrewpbrett) [inline engines/order_management/spec/services/order_management/stock/packer_spec.rb:13]: Why did we need to change this?

**Review by andrewpbrett — APPROVED** [review]
This all looks great! I think it's ready for careful manual testing. Apart from playing around with the Gemfile I didn't actually open up the site locally with this branch, just based it on what I saw. I'd be happy to help the testing team do so, though, it will be a big job. 

The biggest things I'd be careful to test based on what I saw in these changes are invoices and reports (as noted already in the issue). 

A lot of the comments I left don't necessarily require a response, they're more for my own curiosity. 

Nice work!! 🎉

File: Gemfile
Comment (sauloperez) [inline Gemfile:94]: the end of an era :joy: :wave:

File: app/models/enterprise.rb
Comment (sauloperez) [inline app/models/enterprise.rb:135]: I believe `where(enterprises: { is_primary_producer: true })` would work as well FYI. It doesn't make much difference here so I don't mean to change it.

File: app/services/invoice_renderer.rb
Comment (sauloperez) [inline app/services/invoice_renderer.rb:3]: I would very much prefer `renderer.order = order`. These things tend to end badly.

File: config/initializers/paranoia.rb
Comment (sauloperez) [reply config/initializers/paranoia.rb:1]: I'd deal with it after successfully upgrading. It'll definitely make the next upgrade harder, at least.

File: lib/open_food_network/customers_report.rb
Comment (sauloperez) [inline lib/open_food_network/customers_report.rb:64]: it should be possible to do `where(spree_products: { supplier_id: params[:supplier_id].to_i)` as well, but again, just a minor comment.

File: app/models/enterprise.rb
Comment (Matt-Yorkley) [reply app/models/enterprise.rb:135]: I don't remember the exact details, but i think there were some really nasty Postgres errors around this. I possibly explain it in the commit message...

File: app/services/invoice_renderer.rb
Comment (Matt-Yorkley) [reply app/services/invoice_renderer.rb:3]: Agreed! I tried various nicer fixes in this case, but that solution would require a massive rewrite of lots of other connected bits of code, which also got increasingly messy.

Comment (Matt-Yorkley) [reply app/services/invoice_renderer.rb:3]: The whole implementation of this `InvoiceRenderer` is not great.

File: app/models/spree/preferences/store.rb
Comment (Matt-Yorkley) [reply app/models/spree/preferences/store.rb:64]: It was something to do with `Spree::Preferences` passing `nil` as the key here, but only in some rare cases. It subsequently passes it to `Rails.cache.delete`, and `nil` is not an acceptable argument to that method, so it throws a fatal error. I didn't get much deeper into the specifics of the underlying cause, maybe the `#delete` method changed slightly?

File: config/environments/production.rb
Comment (Matt-Yorkley) [reply config/environments/production.rb:47]: I'll have to look at #6178. The change removed a deprecation message in previous versions, but it might even be hard-deprecated in 4.2...

File: spec/models/enterprise_caching_spec.rb
Comment (Matt-Yorkley) [reply spec/models/enterprise_caching_spec.rb:27]: I think it's unrelated to that, but possibly some change in behavior in the `#change` method? I'm not entirely sure. All these specs were 100% broken, but they're actually just testing whether or not `#touch` works when saving a record that is supposed to change the `updated_at` value of connected records, which isn't even caching behavior (although it's important for some of our cache invalidation to work).

It seems like the object was not being reloaded correctly between the before and after checks as part of the #change` matcher. Eg: the object actually _had_ changed but the test wasn't seeing it.

:man_shrugging:

File: spec/services/bulk_invoice_service_spec.rb
Comment (Matt-Yorkley) [reply spec/services/bulk_invoice_service_spec.rb:12]: There's still an expect here at the start of the block (above). We're expecting the service to initiate a `BulkInvoiceJob`, and it does. The logic in the previous generic delayed job processing with `#start_pdf_job_without_delay` has moved location into `BulkInvoiceJob`.

File: spec/features/consumer/authentication_spec.rb
Comment (Matt-Yorkley) [reply spec/features/consumer/authentication_spec.rb:107]: We're still checking that clicking the "reset password" button sends an email. We're not diving into the specifics of what that email looks like, but I think it might be sufficient in this case. The interface for examining some of these things in tests changed substantially and in some cases it was more difficult to check the specifics.

File: lib/open_food_network/payments_report.rb
Comment (Matt-Yorkley) [reply lib/open_food_network/payments_report.rb:114]: Looking back at the change I think that was part of a Rubocop autocorrect...?

But I would say that space shouldn't be there...

File: app/services/invoice_renderer.rb
Comment (Matt-Yorkley) [reply app/services/invoice_renderer.rb:3]: I just had another look at this, and confirmed we get broken specs with `renderer.order = order`. 

```
NoMethodError: undefined method `order='
```

I can't tell you how nasty this was to debug and fix :see_no_evil:

File: spec/features/consumer/authentication_spec.rb
Comment (Matt-Yorkley) [reply spec/features/consumer/authentication_spec.rb:107]: I added a commit :+1:

File: lib/open_food_network/payments_report.rb
Comment (Matt-Yorkley) [reply lib/open_food_network/payments_report.rb]: I added a commit to remove the two strange-looking spaces in this file :+1:

File: spec/controllers/admin/bulk_line_items_controller_spec.rb
Comment (Matt-Yorkley) [reply spec/controllers/admin/bulk_line_items_controller_spec.rb:198]: I guess so. I don't know why it would have rendered a string with a space in it before, but it definietly renders an empty string now... :man_shrugging:

File: config/environments/production.rb
Comment (Matt-Yorkley) [reply config/environments/production.rb:47]: #6178 seems fine :+1:

File: config/initializers/paranoia.rb
Comment (Matt-Yorkley) [reply config/initializers/paranoia.rb:1]: I think this bug only affects the test suite. It could be that we can remove this after we merge this branch and upgrade Capybara / Selenium, but we'll have to see.

File: app/services/invoice_renderer.rb
Comment (sauloperez) [reply app/services/invoice_renderer.rb:3]: fair enough. Getting the upgrade out << instance_variable_set. Could you please create an issue for it so this fact doesn't get lost?

File: engines/order_management/spec/services/order_management/stock/packer_spec.rb
Comment (Matt-Yorkley) [reply engines/order_management/spec/services/order_management/stock/packer_spec.rb:13]: I've just had a look at this again and done some debugging. Setting the variant's weight in this way is actually not how variant weight is set, it's a computed value. I tried adding some simpler cases where a variant's weight is updated this way, and they failed. I tried the same tests in master and they also failed. Eg:

```
variant.weight.update(weight: 1)
expect(variant.reload.weight).to eq 1
```

The test fails and result is zero in both this branch and master.

Updating a variant's weight correctly is done via `variant.update(unit_value: 1)`. I'll add a commit now.

Comment (andrewpbrett) [reply engines/order_management/spec/services/order_management/stock/packer_spec.rb]: Ahhh nice sleuthing! 🎉

**Review by luisramos0 — APPROVED** [review]
Test Ready?

I think you can remove the application_responder.rb comment, create the issue Pau asked for (for instance_variable_set), maybe create another issue to upgrade money back to 5.1.1 and bang, Test Ready :heart_eyes:

**Comment by Matt-Yorkley** [comment]
- Comment removed :+1: 
- InvoiceRenderer issue: https://github.com/openfoodfoundation/openfoodnetwork/issues/6410
- Spree::Money issue: https://github.com/openfoodfoundation/openfoodnetwork/issues/6401

**Comment by sauloperez** [comment]
Sorry, I couldn't help moving this to test ready before finishing the build :sweat_smile:

**Comment by luisramos0** [comment]
This is going to be a great xmas gift :-)))

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,
I ran a first smoke test on this branch and it looks pretty good! :muscle: I checked previous shops and created new ones. All good in what setting up users/shops/products and their settings is concerned - more details to come.
However, there is something not quite right on check-out: opened issue #6411 Thought I'd give an early heads-up :+1:

File: app/helpers/checkout_helper.rb
Comment (luisramos0) [inline app/helpers/checkout_helper.rb:22]: At some point we discussed this and we went for to_a.sum instead of map. Maybe not a big difference? 
It would be good to have a unit test for this bug.

Comment (sauloperez) [reply app/helpers/checkout_helper.rb:22]: I also I checked this. I bet there's not much difference but the actual issue is that `enterprise_fee_adjustments` is an array instead of an AR::Relation at this point already and we rely on Ruby to do the DB's work :see_no_evil:

**Comment by Matt-Yorkley** [comment]
Rebased both branches to deal with merge conflicts.

**Review by luisramos0 — APPROVED** [review]
:+1:

**Pull request merged**
